gpbft/gpbft.go

```diff
@@ -940,33 +939,35 @@ func (i *instance) rebroadcastTimeoutElapsed() bool {
 }
 
 func (i *instance) rebroadcast() error {
-
-	// Rebroadcast all messages from the current and previous rounds, unless the
+	// Rebroadcast quality and all messages from the current and previous rounds, unless the
 	// instance has progressed to DECIDE phase. In which case, only DECIDE message is
 	// rebroadcasted.
 	//
 	// Note that the implementation here rebroadcasts more messages than FIP-0086
 	// strictly requires. Because, the cost of rebroadcasting additional messages is
 	// small compared to the reduction in need for rebroadcast.
-	//
-	// It is also not strictly required to sort the rebroadcasted messages. But we do
-	// so for deterministic behaviour, useful for testing.
-	var rounds []uint64
-	for round := range i.broadcasted.messagesByRound {
-		rounds = append(rounds, round)
-	}
-	slices.Sort(rounds)
-	for _, round := range rounds {
-		mbs := i.broadcasted.messagesByRound[round]
-		for _, mb := range mbs {
-			if err := i.participant.host.RequestBroadcast(mb); err != nil {
-				// Silently log the error and proceed. This is consistent with the behaviour of
-				// instance for regular broadcasts.
-				i.log("failed to request rebroadcast %s at round %d: %v", mb.Payload.Phase, mb.Payload.Round, err)
-			} else {
-				i.log("rebroadcasting %s at round %d for value %s", mb.Payload.Phase.String(), mb.Payload.Round, mb.Payload.Value)
-				metrics.reBroadcastCounter.Add(context.TODO(), 1)
-			}
+	var msgs []*MessageBuilder
+	if i.broadcasted.decide != nil {
+		msgs = append(msgs, i.broadcasted.decide)
+	} else {
+		// We rebroadcast messages newest to oldest, except quality which we always
+		// rebroadcast first. That way, if we try to rebroadcast too many at once and some
+		// get dropped, we've already rebroadcast the most useful messages.
+		msgs = append(msgs, i.broadcasted.previousRound...)
+		msgs = append(msgs, i.broadcasted.currentRound...)
+		if i.broadcasted.quality != nil {
+			msgs = append(msgs, i.broadcasted.quality)
+		}
+		slices.Reverse(msgs)
```

Comment:
@masih does this prioritization make sense?

Reply:
> It does, though the pedantic in me would say gossipsub makes no sequential delivery guarantees.

It doesn't but... it drops messages once queues start filling up, so earlier messages are more likely to be delivered (just by how it's implemented).